core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java

```diff
@@ -0,0 +1,154 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Loads {@link ResourcePack}s within a {@link Path} directory, firing the {@link GeyserLoadResourcePacksEvent}.
+ */
+public class ResourcePackLoader implements RegistryLoader<Path, Map<String, ResourcePack>> {
+
+    static final PathMatcher PACK_MATCHER = FileSystems.getDefault().getPathMatcher("glob:**.{zip,mcpack}");
+
+    private static final boolean SHOW_RESOURCE_PACK_LENGTH_WARNING = Boolean.parseBoolean(System.getProperty("Geyser.ShowResourcePackLengthWarning", "true"));
+
+    /**
+     * Loop through the packs directory and locate valid resource pack files
+     */
+    @Override
+    public Map<String, ResourcePack> load(Path directory) {
+        Map<String, ResourcePack> packMap = new HashMap<>();
+
+        if (!Files.exists(directory)) {
+            try {
+                Files.createDirectory(directory);
+            } catch (IOException e) {
+                GeyserImpl.getInstance().getLogger().error("Could not create packs directory", e);
+            }
+
+            return packMap; // As we just created the directory it will be empty
+        }
+
+        List<Path> resourcePacks;
+        try (Stream<Path> stream = Files.walk(directory)) {
+            resourcePacks = stream.filter(PACK_MATCHER::matches)
+                    .collect(Collectors.toCollection(ArrayList::new)); // toList() does not guarantee mutability
+        } catch (Exception e) {
+            GeyserImpl.getInstance().getLogger().error("Could not list packs directory", e);
+            return packMap;
+        }
+
+        GeyserLoadResourcePacksEvent event = new GeyserLoadResourcePacksEvent(resourcePacks);
+        GeyserImpl.getInstance().eventBus().fire(event);
+
+        for (Path path : event.resourcePacks()) {
+            try {
+                GeyserResourcePack pack = readPack(path);
+                packMap.put(pack.manifest().header().uuid().toString(), pack);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        return packMap;
+    }
+
+    /**
+     * Reads a resource pack at the given file. Also searches for a file in the same directory, with the same name
+     * but suffixed by ".key", containing the content key. If such file does not exist, no content key is stored.
+     *
+     * @param path the file to read from, in ZIP format
+     * @return a {@link ResourcePack} representation
+     * @throws IllegalArgumentException if the pack manifest was invalid or there was any processing exception
+     */
+    public static GeyserResourcePack readPack(Path path) throws IllegalArgumentException {
+        AtomicReference<GeyserResourcePackManifest> manifestReference = new AtomicReference<>();
+        GeyserResourcePack pack;
```

Comment:
Remove this variable as the pack is returned in only one location. Apart from that everything seems alright to me

The ResourcePack/Header/Manifest API and what it will look like in the future just needs to be considered by Redned

Reply:
resolved; thanks